mmv1/third_party/terraform/functions/region_from_zone.go

```diff
@@ -0,0 +1,51 @@
+package functions
+
+import (
+	"context"
+
+	"github.com/hashicorp/terraform-plugin-framework/function"
+)
+
+var _ function.Function = RegionFromZoneFunction{}
+
+func NewRegionFromZoneFunction() function.Function {
+	return &RegionFromZoneFunction{}
+}
+
+type RegionFromZoneFunction struct{}
+
+func (f RegionFromZoneFunction) Metadata(ctx context.Context, req function.MetadataRequest, resp *function.MetadataResponse) {
+	resp.Name = "region_from_zone"
+}
+
+func (f RegionFromZoneFunction) Definition(ctx context.Context, req function.DefinitionRequest, resp *function.DefinitionResponse) {
+	resp.Definition = function.Definition{
+		Summary:     "Returns the region within a provided resource's zone",
+		Description: "Takes a single string argument, which should be a resource's zone.",
+		Parameters: []function.Parameter{
+			function.StringParameter{
+				Name:        "zone",
+				Description: "A string of a resource's zone.",
+			},
+		},
+		Return: function.StringReturn{},
+	}
+}
+
+func (f RegionFromZoneFunction) Run(ctx context.Context, req function.RunRequest, resp *function.RunResponse) {
+	// Load arguments from function call
+	var arg0 string
+	resp.Diagnostics.Append(req.Arguments.GetArgument(ctx, 0, &arg0)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Validate input - can't be an empty string, every zone will have 2nd to last element in string be a `-`
+	if len(arg0) != 0 && arg0[len(arg0)-2] == '-' {
+		resp.Diagnostics.Append(resp.Result.Set(ctx, arg0[:len(arg0)-2])...)
+	} else {
+		return
+	}
```

Comment:
This code is doing both input validation and setting the result in one step, which makes it harder to parse.

In your logic for this function could you please:
- first implement checks for good input, and handle setting any warnings or errors to provide feedback to users
    - If the input is determined to be 'bad' you would then return early
        - [Read here about the "return early pattern"](https://medium.com/swlh/return-early-pattern-3d18a41bba8) if you've not encountered this idea before! 
    - [Here's an example](https://github.com/GoogleCloudPlatform/magic-modules/blob/431dd77ab017e25354dd7309ef792858ff2a6d78/mmv1/third_party/terraform/functions/element_from_id.go#L14-L34) of how you can set errors and warnings in provider functions
- second, implement the code for setting and returning the result
    - if you've returned early prior to this step you can assume the string is valid, so the code will just be `resp.Diagnostics.Append(resp.Result.Set(ctx, arg0[:len(arg0)-2])...)`

Reply:
tests have been added and are passing, another internal test could be added that does a more thorough check on the zone input string. Could use a regex that checks that the first word before the first `-` is a valid location (us,africa,europe,asia, etc.)

Though this could be unnecessary, thoughts?

Reply:
I think this would be a problem in the long-run as GCP adds new locations. The validation would fall out of sync with GCP and be a pain to maintain.